### PR TITLE
Consume XAML straddle headers via nuget 

### PR DESCRIPTION
**Pull request by asklar**
This removes the headers that are used for straddling system XAML and WinUI 3, and moves to instead depend on a nuget package.
This will do a few things:
- Enable code reuse of the headers by other projects that need to dual-target (like telerik etc)
- Make the dependency on UI framework explicit in the projects that depend on it.
- Make it easier to clean up once we move to winui 3 (i.e. make what needs to be cleaned up more obvious)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5606)

### CODE REVIEW

**Comment by NickGerleman** [comment]
Definitely a fan of removing the conditional headers in one way or another. I added them as a quick solution to our bloated PCH issue, but know they're not particularly clean.

Found the repo for this at https://github.com/asklar/XAMLStraddle

A couple of quick questions/comments:
1. I noticed the package is a copy of the headers we currently use in RNW, instead of the full API surface. We've seen some PRs that have needed to add new headers. What would this workflow look like with multiple repos?
2. There have been compliance concerns in the past when developers have wanted to use NuGet packages derived from code in their own private repo rather than an msft one (ACoates will know more). Do you know if we run into any issues for this case?

Possibly out of scope, but if we see other projects with the same requirements, it would be great for WinUI itself to document a facility or headers like this.

**Comment by asklar** [comment]
> Definitely a fan of removing the conditional headers in one way or another. I added them as a quick solution to our bloated PCH issue, but know they're not particularly clean.
> 
> Found the repo for this at https://github.com/asklar/XAMLStraddle
> 
> A couple of quick questions/comments:
> 
> 1. I noticed the package is a copy of the headers we currently use in RNW, instead of the full API surface. We've seen some PRs that have needed to add new headers. What would this workflow look like with multiple repos?
> 2. There have been compliance concerns in the past when developers have wanted to use NuGet packages derived from code in their own private repo rather than an msft one (ACoates will know more). Do you know if we run into any issues for this case?
> 
> Possibly out of scope, but if we see other projects with the same requirements, it would be great for WinUI itself to document a facility or headers like this.

- New headers or changes to the headers can easily be made by checking into the straddle repo, and publishing a new nuget, then retargetting the RNW code to use this new nuget
- Fully agree on the sentiment that WinUI should support something like this out of the box, I don't see it happening in the short term.

**Comment by acoates-ms** [comment]
If it's just source consumption, then it should be fine for compliance.

I'm more worried about repo maintenance. -- We should make sure that the repo in question has enough permissions to go around.  Can non-Microsoft projects use Microsoft GitHub teams in the permissions?  Can you add react-native-windows-write as having write permission to the new repo?  Does it auto publish new versions of the NuGet?  Who has permission to publish the NuGet?

**Comment by dannyvv** [comment]
Re NuGet publish: I just went through the steps on: https://aka.ms/nuget to support our publish to nuget.org. 
> Microsoft packages should only be owned by a group/team account that publishes the package and the Microsoft account.

File: vnext/Desktop/React.Windows.Desktop.vcxproj [hunk 1]
```diff
@@ -325,6 +325,7 @@
     <Import Project="$(SolutionDir)\packages\Microsoft.Build.Tasks.Git.1.0.0\build\Microsoft.Build.Tasks.Git.targets" Condition="Exists('$(SolutionDir)\packages\Microsoft.Build.Tasks.Git.1.0.0\build\Microsoft.Build.Tasks.Git.targets')" />
     <Import Project="$(SolutionDir)\packages\Microsoft.SourceLink.Common.1.0.0\build\Microsoft.SourceLink.Common.targets" Condition="Exists('$(SolutionDir)\packages\Microsoft.SourceLink.Common.1.0.0\build\Microsoft.SourceLink.Common.targets')" />
     <Import Project="$(SolutionDir)\packages\Microsoft.SourceLink.GitHub.1.0.0\build\Microsoft.SourceLink.GitHub.targets" Condition="Exists('$(SolutionDir)\packages\Microsoft.SourceLink.GitHub.1.0.0\build\Microsoft.SourceLink.GitHub.targets')" />
+    <Import Project="$(SolutionDir)\packages\XamlStraddle.0.0.3\build\XamlStraddle.targets" Condition="Exists('$(SolutionDir)\packages\XamlStraddle.0.0.3\build\XamlStraddle.targets')" />
```
Comment (vmoroz) [inline vnext/Desktop/React.Windows.Desktop.vcxproj:328]: >Import [](start = 5, length = 6)

The desktop project must not depend on XAML.

File: vnext/Desktop.ABITests/packages.config [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-	<package id="Microsoft.googletest.v140.windesktop.msvcstl.static.rt-dyn" version="1.8.1" targetFramework="native" />
-	<package id="Microsoft.Windows.CppWinRT" version="2.0.200615.7" targetFramework="native" />
+  <package id="Microsoft.googletest.v140.windesktop.msvcstl.static.rt-dyn" version="1.8.1" targetFramework="native" />
+  <package id="Microsoft.Windows.CppWinRT" version="2.0.200615.7" targetFramework="native" />
+  <package id="XamlStraddle" version="0.0.3" targetFramework="native" />
```
Comment (vmoroz) [inline vnext/Desktop.ABITests/packages.config:5]: >XamlStraddle [](start = 15, length = 12)

The ABITests must not depend on XAML.

File: vnext/Desktop/React.Windows.Desktop.vcxproj (hunk 1 above)
Comment (asklar) [reply vnext/Desktop/React.Windows.Desktop.vcxproj:328]: Agree but it does today. Feel free to file/fix though :-)

File: vnext/Desktop.ABITests/packages.config (hunk 2 above)
Comment (asklar) [reply vnext/Desktop.ABITests/packages.config:5]: Agree but it does today. Feel free to file/fix though :-)

**Comment by kmelmon** [comment]
Overall I like the code sharing.  Just one question about why the repo is apparently a private repo instead of a msft repo.  This can be changed as a follow-up if necessary.

---
In reply to: [666398353](https://github.com/microsoft/react-native-windows/pull/5606#issuecomment-666398353) [](ancestors = 666398353)

**Comment by NickGerleman** [comment]
Setting up a new MSFT repo is kind of a pain. Has there been consideration for keeping this in RNW, but as a separate NuGet package that our repo publishes? That would eliminate some of the hassle of multiple repos, but still allow consumption outside of the repo and create a clear separation.

File: packages/@react-native-windows/cli/src/generator-windows/index.ts [hunk 3]
```diff
@@ -181,12 +186,20 @@ export async function copyProjectTemplateAndReplace(
...
       hasProps: true,
       hasTargets: true,
+      framework: 'native',
     },
     {
       id: options.useWinUI3 ? 'Microsoft.WinUI' : 'Microsoft.UI.Xaml',
       version: options.useWinUI3 ? winui3Version : '2.3.191129002',
       hasProps: false, // WinUI/MUX props and targets get handled by RNW's WinUI.props.
       hasTargets: false,
+      framework: 'native',
```
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/generator-windows/index.ts:196]: Is this ever not native for our templates?

**Pull request closed without merging**